File: app/ionic/src/server/Interfaces/Propers.ts
//
//  Propers.ts
//  ordo-62
//
//  Created by Matthew Frankland on 29/01/2024.
//

class PropersData {
  private title?: string
  public english: string
  public latin: string
  private prayers: { [ id: string ] : string }

  constructor ( json: string ) {
    let obj = JSON.parse ( json )
    this.english = obj.english
    this.latin = obj.latin
    this.prayers = { "English": obj.english, "Latin": obj.latin }
    if ( obj.title ) {
      this.title = obj.title
    }
  }

  public GetTitle ( ): string | undefined {
    return this.title
  }

  public GetPrayer ( lang: string ): string {
    if ( lang in this.prayers ) {
      return this.prayers [ lang ]
    }
    return ""
  }

}

File: client/app/xcode/ios/Server/Models/Propers.swift
//
//  Propers.swift
//  ordo-1962
//
//  Created by Matthew Frankland on 15/12/2023.
//

import Foundation

struct PropersData: Codable {
    private let title: String?
    public let english: String
    public let latin: String
    private var prayers: [ String: String ]
    
    init ( title: String?, english: String, latin: String ) {
        self.title = title
        self.english = english
        self.latin = latin
        self.prayers = [ "English": english, "Latin": latin ]
    }
    
    init ( from decoder: Decoder ) throws {
        let values = try decoder.container ( keyedBy: CodingKeys.self )
        self.title = try values.decode ( String.self, forKey: .title )
        self.english = try values.decode ( String.self, forKey: .english )
        self.latin = try values.decode ( String.self, forKey: .latin )
        self.prayers = [
            "English": try values.decode ( String.self, forKey: .english ),
            "Latin": try values.decode ( String.self, forKey: .latin )
        ]
    }
    
    enum CodingKeys: String, CodingKey {
        case title, english, latin
    }
    
    func GetTitle ( ) -> String? {
        return self.title
    }
    
    func GetPrayer ( lang: String ) -> String {
        if let prayer = self.prayers [ lang ] {
            return prayer
        }
        return ""
    }
}
